andersoncd/solver.py

```diff
@@ -246,21 +244,22 @@ def solver(
 
 
 @njit
-def _kkt_violation(w, X, R, penalty, ws):
-    n_samples = X.shape[0]
+def _kkt_violation(w, X, Xw, datafit, penalty, ws):
     grad = np.zeros(ws.shape[0])
     for idx, j in enumerate(ws):
-        grad[idx] = X[:, j] @ R / n_samples
+        grad[idx] = datafit.gradient_scalar(X, w, Xw, j)  # TODO
     return penalty.subdiff_distance(w, grad, ws)
 
 
 @njit
-def _cd_epoch(X, w, R, penalty, lc, feats):
-    n_samples = R.shape[0]
+def _cd_epoch(X, w, Xw, datafit, penalty, lc, feats):
+    n_samples = Xw.shape[0]
     for j in feats:
         Xj = X[:, j]
         old_w_j = w[j]
         w[j] = penalty.prox_1d(
-            old_w_j + Xj @ R / lc[j], n_samples / lc[j], j)
+            old_w_j + datafit.gradient_scalar(
+                X, w, Xw, j) / lc[j], n_samples / lc[j], j)
+        # TODO use datafit.grad_j and datafit.lipschitz
         if w[j] != old_w_j:
-            R += (old_w_j - w[j]) * Xj
+            Xw -= (old_w_j - w[j]) * Xj
```

Comment:
This change leads to overflow, and I do ont understand why

Reply:
I think this  is good now

Reply:
yes :)